### PR TITLE
Features/22 percentile

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

Introducing statistics.percentile(), along the numpy.percentile() lines.

```
Compute the q-th percentile of the data along the specified axis.
Returns the q-th percentile(s) of the tensor elements.

```

Also implemented statistics.median() --> percentile(q=50).

Issue/s resolved: #22, #175 

### Strategy
Working assumptions: 
- `q` is always going to be a smallish (1-D) DNDarray. It must reside on all ranks
- the result of percentile() needs to be available on all processes --> non-distributed DNDarray.
- because percentile works on sorted data, percentiles are calculated locally and Bcasted if necessary

General procedure:
1. sort data along axis 
2. calculate indices matching q values based on interpolation
3. run local_percentile on sorted data (pass indices)
   - in local_percentile: 
         - if indices are int, percentile is the slice of data matching those indices **along axis**
         - if indices are float, percentile is as above + weighted difference between data at indices and data at indices+1. Weights as per interpolation method (midpoint or linear).
         - permute dimensions 0 <--> axis if axis != 0 
           (cf. https://numpy.org/devdocs/reference/generated/numpy.percentile.html)

If data are distributed EDIT: and `split == axis`
1., 2. as above
3. in order to Bcast from the correct root later, we need a map of what indices (i.e., what q values) will be calculated on what rank (or at least, that's the solution I came up with). The map is `indices_map(size, nperc)` (size=number of ranks, nperc=number of percentiles/elements in q). indices_map[m][n] stores index[n] corresponding to q[n] that will be calculated on rank m. All other indices_map values (no q on rank) are -1. (~~In fact, we only need this if axis=split. I will fix it later.~~ FIXED)
4. we also need to know what dimension the various bits of percentile will be joined along. This is the `join` axis
5. allocate memory for percentile tensor on all ranks.
6. based on indices_map, if rank contains percentile-relevant data, run local_percentile as above
7. Bcast into appropriate slice of percentile from appropriate root

## Changes proposed:
- ht.percentile(x, q) returns a non-distributed heat tensor
- available interpolations: linear, lower, higher, midpoint, nearest (default: linear)
- Known issues: 
   - cannot handle empty processes for now (#568 ); 
   - axis can only be int, tuple of ints not implemented yet
   - out buffer not implemented yet 
   - keepdim=True, if axis is None the result does not broadcast against the original array x

## Type of change

- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/statistics.py [hunk 1]
```diff
@@ -1100,6 +1132,244 @@ def mpi_argmin(a, b, _):
...
+                percentile[perc_slice].shape, dtype=percentile.dtype, comm=x.comm
+            )
+            proc = perc_map[i][1]
+            if indices.numel() and rank == proc:
+                local_p = factories.array(local_percentile(data, axis, indices))
+            x.comm.Bcast(local_p, root=proc)
+            percentile[perc_slice] = local_p
+    else:
+        local_p = factories.array(local_percentile(data, axis, indices))
+        percentile = local_p
```
Comment (coquelin77) [inline heat/core/statistics.py:1365]: can these be combined?

File: heat/core/tests/test_statistics.py [hunk 2]
```diff
@@ -888,6 +890,67 @@ def test_minimum(self):
         with self.assertRaises(ValueError):
             ht.minimum(random_volume_1, random_volume_2, out=output)
 
+    def test_percentile(self):
```
Comment (coquelin77) [inline heat/core/tests/test_statistics.py:893]: can you add a different dtype to the tests?

File: heat/core/statistics.py [hunk 3]
```diff
@@ -1100,6 +1132,231 @@ def mpi_argmin(a, b, _):
...
+
+    # edge-case: x is a scalar. Return x
+    if x.numdims == 0:
+        percentile = x._DNDarray__array * q / q
+        return factories.array(percentile, split=None, dtype=x.dtype, device=x.device, comm=x.comm)
+
+    # no axis
+    if axis is None:
+        if x.numdims > 1:
+            x = x.copy()
```
Comment (Cdebus) [inline heat/core/statistics.py:1272]: Puh, this could easily run into memory problems for large tensors, no?

**Comment by Cdebus** [comment]
@coquelin77 is right, could you give a short overview on how the workflow is?

File: heat/core/statistics.py [hunk 4]
```diff
@@ -1100,6 +1132,231 @@ def mpi_argmin(a, b, _):
...
+    With this option, the result can broadcast correctly against the original array a.
+
+    Returns:
+    --------
+
+    percentile : process-local ht.tensor
+
+    """
+
+    def local_percentile(data, axis, indices):
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1174]: > @coquelin77 is right, could you give a short overview on how the workflow is?

Thanks @coquelin77 , @Cdebus , I've added a "strategy" section to the general description.

**Comment by ClaudiaComito** [comment]
> can you give a basic outline of the strategy here? I think i understant what you are doing but i want to make sure. also, should the end be balanced?

Thanks @coquelin77 , you're right, I expanded the general description. The end should be split=None

File: heat/core/statistics.py [hunk 5]
```diff
@@ -1100,6 +1132,244 @@ def mpi_argmin(a, b, _):
...
+def percentile(x, q, axis=None, interpolation="linear", keepdim=False):
+    """
+    Compute the q-th percentile of the data along the specified axis.
+    Returns the q-th percentile(s) of the tensor elements.
+
+    Parameters:
+    -----------
+
+    a : ht.tensor
+    Input tensor
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: Done, thanks

File: heat/core/tests/test_statistics.py (hunk 2 above)
Comment (ClaudiaComito) [reply heat/core/tests/test_statistics.py:893]: Done, as in I left the original data tensor as int. ht.percentile as well as numpy.percentile always returns float64.

File: heat/core/statistics.py [hunk 6]
```diff
@@ -1100,6 +1132,231 @@ def mpi_argmin(a, b, _):
...
+
+        return percentile
+
+    # sanitize input
+    if not isinstance(x, dndarray.DNDarray):
+        raise TypeError("expected x to be a ht.DNDarray, but was {}".format(type(x)))
+    if isinstance(axis, list) or isinstance(axis, tuple):
+        raise NotImplementedError("ht.percentile(), tuple axis not implemented yet")
+
+    if isinstance(q, dndarray.DNDarray):
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1240]: It's an array_like of floats in numpy's implementation. I agree that it most cases it will be passed as a list or tuple of floats. However, one might have a `q = np.array(x,y,z)` that they pass to their numpy percentile calculations, and they'll just swap np with ht. I'd like the option to be available.

File: heat/core/statistics.py [hunk 7]
```diff
@@ -1100,6 +1132,231 @@ def mpi_argmin(a, b, _):
...
+    Compute the q-th percentile of the data along the specified axis.
+    Returns the q-th percentile(s) of the tensor elements.
+
+    Parameters:
+    -----------
+
+    a : ht.tensor
+    Input tensor
+
+    q : ht.tensor, torch.tensor, scalar, or list of scalars
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: Yeah you're right, I wasn't sure about this one, I'll remove it.

File: heat/core/statistics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: I moved the check for tensor q to the last place in the if... elif... series, so in most cases it will just be skipped.

File: heat/core/statistics.py [hunk 8]
```diff
@@ -1100,6 +1132,231 @@ def mpi_argmin(a, b, _):
...
+
+    """
+
+    def local_percentile(data, axis, indices):
+        """
+        Process-local percentile calculations
+
+        Input:
+        ------
+        data : ht.tensor
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: I need the halo... but you're right, it should be torch-in, torch-out. I'll fix this

File: heat/core/statistics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1174]: It's the same procedure whether the data are distributed or not. I find it easier to follow this way. What would be a better way to do it? We also have local_min and local_max for the min() and max() implementations?

File: heat/core/statistics.py (hunk 3 above)
Comment (ClaudiaComito) [reply heat/core/statistics.py:1271]: YES.

Comment (Cdebus) [reply heat/core/statistics.py:1266]: Especially since you later copy x into data for sorting.... that I get, there is no way around it. But considering this you would make 2 extra copies..... Maybe think about fusing this for the axis=None case?

File: heat/core/statistics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Cdebus) [reply heat/core/statistics.py]: I see :) Took me a while to understand the workflow

File: heat/core/statistics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Cdebus) [reply heat/core/statistics.py]: Well actually there is a lot of torch functionality that could be easily  moved to heat.DNDarrays I think

File: heat/core/statistics.py [hunk 9]
```diff
@@ -1100,6 +1132,239 @@ def mpi_argmin(a, b, _):
...
+        q = q._DNDarray__array
+    else:
+        raise TypeError(
+            "ht.DNDarray, torch.tensor, list or tuple supported, but q was {}".format(type(q))
+        )
+    nperc = q.numel()
+
+    # edge-case: x is a scalar. Return x
+    if x.numdims == 0:
+        percentile = x._DNDarray__array * q / q
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1260]: Well, I'm multiplying by "ones", as many ones as quantiles.

[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py:1260]: So, this is equivalent to:

`percentile = x._DNDarray__array * torch.ones(nperc, dtype=..., device=...)`

File: heat/core/statistics.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: > You use a lot of torch.Tensors, which is not necessary as far as I see. Would be better to transfer this functionality to ht.DNDarrays as far as possible and call _DNDarray__array if the local torch tensor is required

Hmm, I can see that it might make things easier to follow when reading the code. But there are lots of instances where we would have to wrap a torch tensor into a DNDarray, just to use the underlying _DNDarray__array for the rest of the function. Why do that?

It would be good to settle this so we have a project-wide agreement. We could have a naming convention so one can tell torch and heat tensors apart at first glance.

**Comment by ClaudiaComito** [comment]
I'm reopening this after a bit of cleaning up. In particular, most (all?) unnecessary  communication for the distributed case when `split != axis` is gone.

File: heat/core/statistics.py [hunk 10]
```diff
@@ -1103,6 +1135,236 @@ def mpi_argmin(a, b, _):
...
+        axis : int or TODO tuple of ints
+        indices : torch.tensor
+
+        Returns:
+        --------
+        percentile, torch.tensor, process-local
+        """
+
+        axis_slice = data.ndim * (slice(None, None, None),)
+        if indices.dtype == torch.long or indices.dtype == torch.int:
```
Comment (Markus-Goetz) [inline heat/core/statistics.py:1193]: Better check here for identity (faster) with the 'is' operator

File: heat/core/statistics.py [hunk 11]
```diff
@@ -1103,6 +1135,236 @@ def mpi_argmin(a, b, _):
...
+            # interpolation 'lower', 'higher', or 'nearest'
+            axis_slice = axis_slice[:axis] + (indices.tolist(),) + axis_slice[axis + 1 :]
+            percentile = data[axis_slice]
+        else:
+            floor_indices = torch.floor(indices).type(torch.long)
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/statistics.py]: @Markus-Goetz good point, it doesn't seem to be necessary, I'm casting all integer indices to torch.int now.

File: heat/core/statistics.py (hunk 10 above)
Comment (ClaudiaComito) [reply heat/core/statistics.py]: Done!

**Comment by Markus-Goetz** [comment]
What is the state on this? Ready for a second review after all changes?

**Comment by ClaudiaComito** [comment]
> What is the state on this? Ready for a second review after all changes?

Yep, go ahead, I just merged master.
